### ISSUE TITLE
pod connectivity failed after node restart

### ISSUE THREAD

**Issue by r0bj** [projectcalico/calico#1173]
Pod connectivity failed after node restart.

Working pod `nginx-ingress-controller-l4-ntjrb`:
```
# kubectl get pods -l k8s-app=nginx-ingress-controller-l4 -n kube-system -o wide
NAME                                READY     STATUS    RESTARTS   AGE       IP             NODE
nginx-ingress-controller-l4-ntjrb   1/1       Running   0          5m        10.203.62.74   dev-k8s-lb-l4-p1
```

```
# calicoctl get workloadEndpoint
[...]
dev-k8s-lb-l4-p1    k8s            kube-system.nginx-ingress-controller-l4-ntjrb          eth0
```

After node `dev-k8s-lb-l4-p1` restart:

```
# kubectl get pods -l k8s-app=nginx-ingress-controller-l4 -n kube-system -o wide
NAME                                READY     STATUS             RESTARTS   AGE       IP        NODE
nginx-ingress-controller-l4-ntjrb   0/1       CrashLoopBackOff   8          22m       <none>    dev-k8s-lb-l4-p1
```

On the node there is no `cali*` interface:
```
# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 52:54:00:95:52:c3 brd ff:ff:ff:ff:ff:ff
    inet 10.14.30.145/24 brd 10.14.30.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::5054:ff:fe95:52c3/64 scope link
       valid_lft forever preferred_lft forever
3: docker0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN group default
    link/ether 02:42:e7:fe:61:a4 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.1/16 scope global docker0
       valid_lft forever preferred_lft forever
4: tunl0@NONE: <NOARP,UP,LOWER_UP> mtu 1440 qdisc noqueue state UNKNOWN group default qlen 1
    link/ipip 0.0.0.0 brd 0.0.0.0
    inet 10.203.62.66/32 scope global tunl0
       valid_lft forever preferred_lft forever
```

Inside pod namespace there is no `eth0` intereface:
```
# docker ps
7146005a40cb        gcr.io/google_containers/pause-amd64:3.0                                                                            "/pause"                 14 minutes ago      Up 14 minutes                           k8s_POD_nginx-ingress-controller-l4-ntjrb_kube-system_c3bba3b4-aa95-11e7-88fd-5254009ef6db_3

# docker inspect --format '{{.State.Pid}}' 7146005a40cb
1579

# nsenter -t 1579 -n ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: tunl0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN group default qlen 1
    link/ipip 0.0.0.0 brd 0.0.0.0
```

There is no workloadEndpoint for `nginx-ingress-controller-l4` pod:
```
# calicoctl get workloadEndpoint | grep nginx-ingress-controller-l4
```

Kubelet logs:
https://gist.github.com/r0bj/baaef22a62575119ede2d6705dd14cdd

calico-node logs:
https://gist.github.com/r0bj/afbfd0929e2a722632fbcfab2c498263

Manually removing pod fixes issue:
```
kubectl delete pod nginx-ingress-controller-l4-ntjrb -n kube-system
```

## Expected Behavior
After node restart pod is recreated and connectivity works.

## Current Behavior
Pod connectivity failed after node restart.

## Your Environment
* Calico version
v2.6.1
* Orchestrator version (e.g. kubernetes, mesos, rkt):
kubernetes 1.7.2
* Operating System and version:
Ubuntu 14.04
docker 1.12.6

**Comment by r7vme** [comment]
we are seeing similar behaviour. After nodes reinstall/reboot IPIP tunnels do not forward traffic. All routes are in place and network between VMs is fine (i can ssh between VMs). In tcpdump i see IPIP traffic from node, but don't see anything on destination node.

What is most weird that as workaround, *i need to ping tunl0 IP addresses. From node A ping node B, from node B ping node A*. After i immediately see ICMP traffic.

I'm seeing this constantly on many enviroments and it not related to AWS, we also use IPIP in out KVM environments and see same behavior there after node reboot/reinstall.

I found that Calico allows only IPIP traffic from known hosts, but this is not the case. I see no matches and see all my hosts in `ipset list`
```
r7vme@ip-10-0-5-117 ~ $ sudo iptables -vL | grep ipencap
    0     0 DROP       ipencap--  any    any     anywhere             anywhere             /* cali:p8Wwvr6qydjU36AQ */ /* Drop IPIP packets from non-Calico hosts */ ! match-set cali4-all-hosts src
```

Any suggestions?

### Details:
Calico 2.5.1 (or 2.3.0)
Kubernetes 1.7.5 (or 1.6.4)
CoreOS 1465.8.0

```
- apiVersion: v1
  kind: ipPool
  metadata:
    cidr: 192.168.0.0/16
  spec:
    ipip:
      enabled: true
      mode: always
    nat-outgoing: true
```

**Comment by heschlie** [comment]
@r0bj I'm having trouble reproducing this, would you be able to provide a bit more detail, I know you've provided a lot to start and thanks for that!

Are you able to reproduce this reliably? What does your Calico yaml look like? What does the deployment for `nginx-ingress-controller-l4` look like? Where are you running your cluster? Are all pods on the restarted node having trouble coming back up if you have more than one?

**Comment by r0bj** [comment]
@heschlie In my opinion this issue is related to timing, notice that:

1. kubelet started at 13:06:44
2. for some reason (more about it later) k8s master address was unavailable until 13:07:27
```
E1006 13:07:26.317926     862 reflector.go:190] k8s.io/kubernetes/pkg/kubelet/kubelet.go:408: Failed to list *v1.Node: Get https://dev.kubernetes-controller.service.consul:6443/api/v1/nodes?fieldSelector=metadata.name%3Ddev-k8s-lb-l4-p1&resourceVersion=0: dial tcp: lookup dev.kubernetes-controller.service.consul on 10.14.30.145:53: read udp 10.14.30.145:41998->10.14.30.145:53: i/o timeout
```
Reason for that was actually issue with local consul agent which was unavailable after host restart for some time. So domain `dev.kubernetes-controller.service.consul` (k8s master) was unable to resolve for that time (but this is another story).
3. calico-node starts at around 13:07:37

After testing different consul options and versions I can say that this is the reason for this issue:

When k8s master domain is available right after kubelet starts everything works as expected.
But when there is delay between kubelet starts and k8s master domain availability, created pods are broken.

calico yaml:
https://gist.github.com/r0bj/9626ac854fffaa4e2adae6b9402cac13

nginx-ingress-controller-l4 yaml:
https://gist.github.com/r0bj/753cdb380869f77de114f104f27118ad

It was easy to reproduce in k8s version 1.7.2.
What's interesting, since I've upgraded k8s to 1.8.1 i couldn't reproduce it any more.

**Comment by heschlie** [comment]
@r0bj Thanks for the extra sleuthing, we're still not able to replicate this so I've got a couple more questions for you if you can.

- Where are you deploying the cluster (AWS, bare metal, GCE)?
- How are you deploying it (by hand, kops, kubeadm, etc.)?
- Are you seeing this problem with all pods on the related node or just the ingress controller?
- When you upgraded to k8s 1.8.1 did you change anything else with the infra?

It is not uncommon for pods to have a couple of restarts while Calico comes up after a Node restart but they should recover almost immediately after Calico is running.

**Comment by r0bj** [comment]
@heschlie 
1. Cluster is deployed on bare metal
2. Cluster is deployed with custom tool based on chef
3. Generally only one pod on node was affected by this issue, nginx-ingress-controller in this case. I don't think it's related but pods affected were mostly DaemonSets.
4. It was just version upgrade, no infra change.

**Comment by namsral** [comment]
I'm having similar issues with IPIP tunneling in a public cloud.

Pod deployments which depend on pods with tunl0 interfaces fail half the time. My current workaround is to deploy a pod on each Kubernetes node which continuously sends a single ICMP packet to tunl0 interfaces to pods are in initialization stage.

I do continuous deployments of Kubernetes clusters with Calico and can reproduce this issue 50% of the time.

My environment:
- Calico version: v2.1.5
- Kubernetes: v1.7.5
- Ubuntu: Ubuntu 16.04 LTS
- Docker: 17.03.1~ce-0~ubuntu-xenial

**Comment by fasaxc** [comment]
@r0bj is it _always DaemonSet pods?  DaemonSet pods bypass the normal scheduler so they can get scheduled before pod networking is ready.  Could be something to do with that.

Are any of your pods host networked?

**Comment by r0bj** [comment]
@fasaxc yes, in my case it were only DaemonSet pods.

I have pods with host network, eg. `kube-proxy`, `node-exporter`,  and of course `calico-node`.

**Comment by fasaxc** [comment]
@r7vme Please can you open a second issue to cover your tunnel issue.  Since it is resolved by pinging down the tunnel, it is likely a different issue to @r0bj's issue, which appears to be an issue with pod re-networking.

**Comment by fasaxc** [comment]
@namsral Please upgrade your Calico to 2.6, we fixed several tunnel issues between 2.1 and 2.6.

**Comment by namsral** [comment]
@fasaxc Will do, thanks.

**Comment by r7vme** [comment]
@fasaxc thanks for taking care.

**Comment by fasaxc** [comment]
I've taken a look at the kubelet logs.  I see this pattern of CNI calls from kubelet to our CNI plugin:

```
PID 1226: kubelet asks us to DEL the interface of pod with a stale container ID, we ignore
PID 1235: kubelet asks us to DEL the interface of pod with a different stale container ID, we ignore
PID 1485: kubelet asks us to DEL the interface of pod with the correct container ID
    PID 1492: IPAM subprocess: succeeds in cleaning up
    Success
PID 2763: kubelet re-sends the same DEL, we log warning that the endpoint doesn't exist
    PID 1492: IPAM subprocess: tries to clean up handle, but is already gone
```

The kubelet never appears to send us an ADD for the new instance of the pod, we only see the above DELs for the old instance.

**Comment by fasaxc** [comment]
AFAICT, given what Kubernetes is sending us, we're doing the right thing.  

This looks like a bug in the kubelet; perhaps its cleanup succeeded (hence all the DELs) but it was unable to report the cleanup to the master (due to the consul issue above) and then it got itself in a broken retry loop, trying to restart a container that it had already cleaned up earlier on:

This gets repeated over and over (but it never seems to call the CNI plugin to try to network the pod):
```
I1006 13:10:23.019748    1797 kuberuntime_manager.go:457] Container {Name:nginx-ingress-controller-l4 Image:gcr.io/google_containers/nginx-ingress-controller:0.9.0-beta.11 Command:[] Args:[/nginx-ingress-controller --configmap=$(POD_NAMESPACE)/nginx-ingress-controller-l4-config --default-backend-service=$(POD_NAMESPACE)/default-http-backend --tcp-services-configmap=$(POD_NAMESPACE)/nginx-ingress-controller-l4-tcp-config --udp-services-configmap=$(POD_NAMESPACE)/nginx-ingress-controller-l4-udp-config] WorkingDir: Ports:[{Name:backend-health HostPort:0 ContainerPort:8080 Protocol:TCP HostIP:} {Name:kube-dns HostPort:0 ContainerPort:5353 Protocol:UDP HostIP:} {Name:nsqlookupd1 HostPort:0 ContainerPort:41601 Protocol:TCP HostIP:} {Name:nsqlookupd2 HostPort:0 ContainerPort:41602 Protocol:TCP HostIP:}] EnvFrom:[] Env:[{Name:POD_NAME Value: ValueFrom:&EnvVarSource{FieldRef:&ObjectFieldSelector{APIVersion:v1,FieldPath:metadata.name,},ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:nil,}} {Name:POD_NAMESPACE Value: ValueFrom:&EnvVarSource{FieldRef:&ObjectFieldSelector{APIVersion:v1,FieldPath:metadata.namespace,},ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:nil,}}] Resources:{Limits:map[] Requests:map[]} VolumeMounts:[{Name:nginx-ingress-serviceaccount-token-x55pz ReadOnly:true MountPath:/var/run/secrets/kubernetes.io/serviceaccount SubPath:}] LivenessProbe:&Probe{Handler:Handler{Exec:nil,HTTPGet:&HTTPGetAction{Path:/healthz,Port:10254,Host:,Scheme:HTTP,HTTPHeaders:[],},TCPSocket:nil,},InitialDelaySeconds:10,TimeoutSeconds:1,PeriodSeconds:10,SuccessThreshold:1,FailureThreshold:3,} ReadinessProbe:&Probe{Handler:Handler{Exec:nil,HTTPGet:&HTTPGetAction{Path:/healthz,Port:10254,Host:,Scheme:HTTP,HTTPHeaders:[],},TCPSocket:nil,},InitialDelaySeconds:10,TimeoutSeconds:1,PeriodSeconds:10,SuccessThreshold:1,FailureThreshold:3,} Lifecycle:nil TerminationMessagePath:/dev/termination-log TerminationMessagePolicy:File ImagePullPolicy:Always SecurityContext:nil Stdin:false StdinOnce:false TTY:false} is dead, but RestartPolicy says that we should restart it.
I1006 13:10:23.019858    1797 kuberuntime_manager.go:741] checking backoff for container "nginx-ingress-controller-l4" in pod "nginx-ingress-controller-l4-ntjrb_kube-system(c3bba3b4-aa95-11e7-88fd-5254009ef6db)"
I1006 13:10:23.019984    1797 kuberuntime_manager.go:751] Back-off 1m20s restarting failed container=nginx-ingress-controller-l4 pod=nginx-ingress-controller-l4-ntjrb_kube-system(c3bba3b4-aa95-11e7-88fd-5254009ef6db)
E1006 13:10:23.020016    1797 pod_workers.go:182] Error syncing pod c3bba3b4-aa95-11e7-88fd-5254009ef6db ("nginx-ingress-controller-l4-ntjrb_kube-system(c3bba3b4-aa95-11e7-88fd-5254009ef6db)"), skipping: failed to "StartContainer" for "nginx-ingress-controller-l4" with CrashLoopBackOff: "Back-off 1m20s restarting failed container=nginx-ingress-controller-l4 pod=nginx-ingress-controller-l4-ntjrb_kube-system(c3bba3b4-aa95-11e7-88fd-5254009ef6db)"
```

@r0bj maybe this was fixed in k8s 1.8?  If you do see it again, I suggest filing a k8s issue.  I'm going to close this but feel free to re-open if you spot a hole in my analysis!

**Comment by fasaxc** [comment]
@namsral Also possible that you were hitting #1357
